openedx/core/djangoapps/course_groups/cohorts.py

```diff
@@ -235,47 +211,85 @@ def get_cohort(user, course_key, assign=True):
         if not assign:
             return None
 
-    choices = course.auto_cohort_groups
-    if len(choices) > 0:
-        # Randomly choose one of the auto_cohort_groups, creating it if needed.
-        group_name = local_random().choice(choices)
+    course_cohorts = get_course_cohorts(course, assignment_type=CourseCohort.RANDOM)
+    if course_cohorts:
+        course_cohort = local_random().choice(course_cohorts)
     else:
-        # Use the "default cohort".
-        group_name = DEFAULT_COHORT_NAME
+        course_cohort, __ = CourseUserGroup.objects.get_or_create(
+            course_id=course_key,
+            group_type=CourseUserGroup.COHORT,
+            name=DEFAULT_COHORT_NAME
+        )
+        CourseCohort.objects.get_or_create(course_user_group=course_cohort, assignment_type=CourseCohort.RANDOM)
+
+    user.course_groups.add(course_cohort)
+
+    return course_cohort
 
-    group, __ = CourseUserGroup.objects.get_or_create(
-        course_id=course_key,
-        group_type=CourseUserGroup.COHORT,
-        name=group_name
-    )
-    user.course_groups.add(group)
-    return group
 
+def migrate_cohort_settings(course):
+    """
+    Migrate all the cohort settings associated with this course from modulestore to mysql.
+    After that we will never touch modulestore for any cohort related settings.
+    """
+    cohort_settings, created = CourseCohortsSettings.objects.get_or_create(
+        course_id=course.location.course_key,
+        defaults={
+            'is_cohorted': course.is_cohorted,
+            'cohorted_discussions': json.dumps(list(course.cohorted_discussions)),
+            'always_cohort_inline_discussions': course.always_cohort_inline_discussions
+        }
+    )
 
-def get_course_cohorts(course):
+    # Add the new and update the existing cohorts
+    if created:
+        # Update the manual cohorts already present in CourseUserGroup
+        manual_cohorts = CourseUserGroup.objects.filter(
```

Comment:
the creation code on the following lines is very similar to that in `add_cohort` - could they be merged together in a `.create` method on the `CourseCohort` model?